### ISSUE TITLE
[REQUEST] Opening links to another website puts container to default

### ISSUE THREAD

**Issue by FormBurden** [mozilla/multi-account-containers#1775]
When I use containers sometimes I use a specific one for only one website (i.e. Microsoft, Youtube, Github, etc.) but when I open up links it still opens up into the same container when the website is different. I would love to see maybe a button that could make it if any website is open that isn't the specific website, it would open into a default container.

**Comment by rotan-hanrahan** [comment]
Totally agree. After finishing on a site and logging out, I am disappointed to see that the selected container stays active even if I click a bookmark, enter a URL into the address bar or perform a search. It can be a few minutes later before I realise that I've let other sites operate within a container that (for example) I've reserved for banking.

**Comment by FormBurden** [comment]
Yeah exactly, I have containers for banking as well. Well, hopefully they add something like this. Would be very useful and MUCH more secure.

**Comment by B00ze64** [comment]
A solution for this is to use Temporary Containers addOn and enable Automatic Mode on Navigation, it will intercept requests to non-assigned sites and open them in a temporary container. Works for me.

**Comment by MikeGitb** [comment]
You can opt to open particular websites always in a particular container. Personally I very much prefer the current behavior to open links in the current container, because - a few exceptions aside - I only use containers to separate broad categories or to manage mutliple different accounts at a single website and more often than not, the links I follow from one site lead me to something that should be bundled to the same topic.

I guess it very much depends on how you are using containers, so it probably should be configurable.

**Comment by rotan-hanrahan** [comment]
+1 for configurability.

I'm now trying the "Temporary Containers" add-on suggested earlier, though it's unnecessarily complicated for the simple use case we're describing here. What I'm looking for is an option to "use the default container when a link in a custom container navigates outside of its container domain(s)". That particular behaviour isn't available in TC as it keeps creating new containers for such links. Essentially, the issue that TC is addressing is not exactly the one described in this ticket, but TC can be forced to give some semblance of protection against unexpected ingression of foreign domains into the current container if you are willing to forego the notion of a "default" catch-all container for such links.

**Comment by FormBurden** [comment]
@MikeGitb Well that's why I suggested a button, kinda like how they have one for "Always open in" feature. Maybe a button below it, just so it doesn't completely change how this great tool works.

**Comment by B00ze64** [comment]
Good day. @rotan-hanrahan,

There are other options I have not tried (below). Temporary Containers is kinda the deluxe version of isolating stuff, it makes sure every website not assigned to a specific container ends-up all alone in own container. The UI for configuration is too complex for its own good, but all you need is "Automatic Mode" and "Isolation on Navigation." You can even use %domain% in the container name so all the temporary containers have the name of the website.

However, there exists more options. For sending non-related domains to the default container you can look at:

1. https://addons.mozilla.org/en-US/firefox/addon/exclusive-containers/
2. https://addons.mozilla.org/en-US/firefox/addon/container-outgoing-links/

Hope this helps.

**Comment by rotan-hanrahan** [comment]
Thanks @B00ze64 for the references.

I'm currently using TC with varying degrees of success but not to the level of a simple "catch-all container" mechanism that I had hoped for.

An interesting problem I encountered with TC was related to the fact that my search provider needed to redirect to a third-party organisation for me to authenticate, and upon success be redirected back. This forced me to create a dedicated container for the search provider and the org. But now when I use search and get a link to a site, clicking that link causes the search page to be replaced by the linked page (in a fresh container), with no "back button". I now middle click to make a new tab instead so I don't lose the search results.

So while the TC add-on is providing some of the secure containment I prefer, there are various "niggles" that are making the UX unpleasant.

The COL add-on appears to have a lingering tab bug, and EC is insufficiently tested in the wild.

By far the best option would be for the core MAC feature to have the option to "open link in default container unless MAC has been configured to open it in a specific container". @MikeGitb has alluded to this already.

Personally I worry about the consequences of endlessly nested add-ons (i.e. add-ons to augment the functionality of other add-ons), and foresee a dependency hell if we're not careful.

**Comment by msanders** [comment]
I would be interested in exclusive containers available as a preference as well. The above referenced [add-on](https://addons.mozilla.org/en-US/firefox/addon/exclusive-containers/) seems to have the close to the behavior I'm looking for, but I agree with @rotan-hanrahan that it would be best to have such a small extension built-in to this one instead. There are also two small bugs with it: the newly opened tab briefly shows the calling container before switching to the new one, and it breaks the behavior of the context item "Open Link in New {Current Container} Tab" (presumably because it works by overloading this function, which is a bit of a hack that could be avoided by having it officially included).

> An interesting problem I encountered with TC was related to the fact that my search provider needed to redirect to a third-party organisation for me to authenticate, and upon success be redirected back.

The behavior I'm looking for is just change the behavior of {middle,cmd,shift}-click on links. When a page opens a new window on its own like this, I'd prefer to keep the existing behavior.

**Comment by B00ze64** [comment]
Hi @rotan-hanrahan,

`An interesting problem I encountered with TC was related to the fact that my search provider needed to redirect to a third-party organisation for me to authenticate, and upon success be redirected back. This forced me to create a dedicated container for the search provider and the org. But now when I use search and get a link to a site, clicking that link causes the search page to be replaced by the linked page (in a fresh container), with no "back button". I now middle click to make a new tab instead so I don't lose the search results.
`

Strange, I have also created a Search Container for Google and al, and when I click on a result, TC opens a new [temporary or permanent] container for me but the original TAB stays there, it does not go away. Since I am VERY used to right-clicking and selecting "Open in a new Tab" this behaviour is very pleasing to me, I no longer have to right-click! Even more so because enabling Containers in general replaces the "Open in new Tab" with "Open in new XX Container Tab" which is pretty much never what I want. There's no BACK of course but it's 1 TAB away to the left lol. If your original Search TAB disappears, something is not quite right. You could open an issue on the TC Github, the guy is quite helpful.

@All, Yeah, MAC should have the option of navigating out of the current container when the domain is not one that is assigned to a container, it is a bit strange that it opens always in the same container. But I suspect this would have to be a setting, because by now lots of people must be depending on that behaviour. If you check [here](https://blog.mozilla.org/tanvi/2017/10/03/update-firefox-containers/) they explain that it was impossible for them to please everyone, which is why we ended-up with addOns helping addOns helping addOns Lol (and TONS of buggy interactions).

Regards,

**Comment by rotan-hanrahan** [comment]
For anyone following this thread, check out the "transition rules" introduced by this fork of MAC: [https://addons.mozilla.org/en-US/firefox/addon/containers-with-transitions/](https://addons.mozilla.org/en-US/firefox/addon/containers-with-transitions/)

Quoting the "About" section:

> "A container transition rule specifies what container to open pages from a given subdomain in depending on the container that they are opened from: so for instance, you could specify that all links to other websites opened from the Facebook container should be opened in a separate Clickbait container by default. You can specify both domain-specific rules and default rules, which are applied when no relevant specific rule exists."

Note that this is a fork, and will not co-reside with the original MAC, and I don't see a way to port a MAC configuration to CWT, so I'm not inclined to try it out (just yet). Would be interested to hear some pro/cons from anyone familiar with it.

**Comment by msanders** [comment]
Looks like this may be addressed by #1654 and #1474.

**Comment by B00ze64** [comment]
@maxxcrawford Yeah, looks like those two might do it, especially container lock #1474. I still do not see in #1654 a way to edit the assigned sites, as-in a text-field where we can copy/paste/edit the site list. It's just impossible to manage a large list of contained sites the way it is right now (and we cannot export the data to another profile either).

**Comment by AdamJel** [comment]
It is possible to set this. You may tick the box 'Limit to Designated Sites' while managing specific container. When ticked, links will open in default container :)

![image](https://user-images.githubusercontent.com/33990009/87778081-eab2f600-c82a-11ea-8f08-30a850dd4b67.png)

**Comment by msanders** [comment]
The above suggestion by @AdamJel has been working great for me so far. Only caveat I've noticed is that I occasionally have to turn it off for sites that do a lot of redirects, which sometimes happen too quickly to add via the "Automatically open this site in..." button. It would be nice if there were a way to edit the site list manually.

**Comment by msanders** [comment]
Alternatively, it would be nice to be able to specify a wildcard for a domain, e.g. `*.site.com`.

**Comment by msanders** [comment]
Relevant issue: https://github.com/mozilla/multi-account-containers/issues/1830.

**Comment by OsmosisD** [comment]
Definitely want this feature (and for it to be toggleable, maybe on a per domain or per container basis).
I use containers to limit tracking and other behaviors. If I open a link from my Facebook or Google container, I want to leave the Facebook or Google specific containers I made so the trackers don't follow.
Basically, "Oh, you're leaving the specified domains for this container, would you like to return to the default/no container?" with Yes/No and "Remember My Choice" or "Add this Domain" options.

**Comment by Decoherent** [comment]
Relevant to [#2152](https://github.com/mozilla/multi-account-containers/issues/2152). Also, hello from the future. This behavior has caused me tremendous difficulty on occasion, like letting my "very personal" container getting accessed by my "Google-box" container. A toggle would be fine; frankly, _anything_ would be fine at this point.
